src/Generation/BuildMethodFragmentGenerator.php

```diff
@@ -0,0 +1,214 @@
+                        continue;
+                    }
+                    $buildMethods = $buildMethods->append(
+                        $this->buildMethodSnippet($methodDetails, $methodSignature, 0 === $i)
+                    );
+                }
+                if ($buildMethods->count() > 0) {
+                    $buildMethodSnippets = $buildMethodSnippets->set(
+                        str_replace('\\', '/', $methodDetails->requestType->getFullname(true)),
+                        $buildMethods
+                    );
+                }
+            }
+        }
+
+        $this->ctx->finalize(null);
+
+        return $buildMethodSnippets;
+    }
+
+    private SourceFileContext $ctx;
+    private ServiceDetails $serviceDetails;
+
+    private function __construct(SourceFileContext $ctx, ServiceDetails $serviceDetails)
+    {
+        $this->ctx = $ctx;
+        $this->serviceDetails = $serviceDetails;
+    }
+
+    public function buildMethodSnippet(MethodDetails $methodDetails, string $methodSignature, bool $isFirst): PhpClassMember
+    {
+        $docType = function ($field): ResolvedType {
+            if ($field->desc->desc->isRepeated()) {
+                if ($field->isEnum) {
+                    return $this->ctx->type(Type::arrayOf(Type::int()), false, true);
+                } elseif ($field->isMap) {
+                    return $this->ctx->type(Type::array());
+                } else {
+                    return $this->ctx->type(Type::arrayOf(
+                        Type::fromField($this->serviceDetails->catalog, $field->desc->desc, false)
+                    ), true, true);
+                }
+            } else {
+                // Affects type hinting for required oneofs.
+                // TODO(vNext) Handle optional oneofs here.
+                if ($field->isEnum) {
+                    return $this->ctx->type(Type::int());
+                } else {
+                    return $this->ctx->type($field->type, true);
+                }
+            }
+        };
+        $docExtra = function ($field): Vector {
+            if ($field->isEnum) {
+                // TODO(vNext): Don't use a fully-qualified name here; and import correctly.
+                $enumType = $field->typeSingular->getFullname();
+                return Vector::new([
+                    "For allowed values, use constants defined on {@see {$enumType}}"
+                ]);
+            } else {
+                return Vector::new([]);
+            }
+        };
+
+        $methodSignatureArguments = array_filter(explode(',', $methodSignature));
+        $requiredFields = $methodDetails->allFields
+            ->filter(fn ($f) => in_array($f->name, $methodSignatureArguments))
+            ->orderBy(fn ($f) => array_search($f->name, $methodSignatureArguments));
+
+        if (count($methodSignatureArguments) !== $requiredFields->count()) {
+            throw new \LogicException('missing method signature arguments');
```

Comment:
Let's make this error more descriptive so its easier to debug e.g. include the method name and signature being evaluated.

Reply:
done!